tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/TraceAnnotations/TraceAnnotationsIntegration.cs

```diff
@@ -0,0 +1,111 @@
+        /// <returns>Calltarget state value</returns>
+        internal static CallTargetState OnMethodBegin<TTarget>(TTarget instance, RuntimeMethodHandle methodHandle, RuntimeTypeHandle typeHandle)
+        {
+            var info = InstrumentedMethodCache.GetOrAdd(
+                    new RuntimeHandleTuple(methodHandle, typeHandle),
+                    key => TraceAnnotationInfoFactory.Create(MethodBase.GetMethodFromHandle(key.MethodHandle, key.TypeHandle)));
+
+            var tags = new TraceAnnotationTags();
+            var scope = Tracer.Instance.StartActiveInternal(info.OperationName, tags: tags);
+            scope.Span.ResourceName = info.ResourceName;
+
+            Tracer.Instance.TracerManager.Telemetry.IntegrationGeneratedSpan(IntegrationId);
+
+            return new CallTargetState(scope);
+        }
+
+        /// <summary>
+        /// OnMethodEnd callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+        /// <param name="state">Calltarget state value</param>
+        /// <returns>A default CallTargetReturn to satisfy the CallTarget contract</returns>
+        internal static CallTargetReturn OnMethodEnd<TTarget>(TTarget instance, Exception exception, in CallTargetState state)
+        {
+            state.Scope.DisposeWithException(exception);
+            return CallTargetReturn.GetDefault();
+        }
+
+        /// <summary>
+        /// OnMethodEnd callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TReturn">Type of the return value</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="returnValue">Return value</param>
+        /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+        /// <param name="state">Calltarget state value</param>
+        /// <returns>A response value, in an async scenario will be T of Task of T</returns>
+        internal static CallTargetReturn<TReturn> OnMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, in CallTargetState state)
+        {
+            // If the return type is Task/Task<T>/ValueTask/ValueTask<TResult>, defer the scope disposal. Otherwise, dispose it now since the async callback will not be invoked
+#if NETCOREAPP3_1 || NET5_0
+            bool closeAsync = returnValue is Task
+                              || returnValue is ValueTask
+                              || (returnValue is not null && returnValue.GetType().IsGenericType && returnValue.GetType().GetGenericTypeDefinition() == typeof(ValueTask<>));
```

Comment:
Do we need to worry about calling returnValue.GetType() twice? Too many years of "reflection is very slow" carved into my brain 😅

Reply:
I was struggling to write it concisely and efficiently, but I think I figured out a way to do it in ac1625fe116deda9664269c8e3ed3eb950a10f7c : `returnValue?.GetType() is Type returnType && returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(ValueTask<>)`